src/audio/mixer.c

```diff
@@ -0,0 +1,646 @@
+	mixer_fx64_t step;
+	/* Length of the waveform (in bytes) */
+	mixer_fx64_t len;
+	/* Length of the loop in the waveform (in bytes) */
+	mixer_fx64_t loop_len;
+	/* Pointer to the waveform */
+	void *ptr;
+	/* Misc flags */
+	uint32_t flags;
+} mixer_channel_t;
+
+typedef struct rsp_mixer_channel_s {
+	uint32_t pos;
+	uint32_t step;
+	uint32_t len;
+	uint32_t loop_len;
+	void *ptr;
+	uint32_t flags;
+} rsp_mixer_channel_t;
+
+_Static_assert(sizeof(rsp_mixer_channel_t) == 6*4);
+
+typedef struct {
+	int max_bits;
+	float max_frequency;
+	int max_buf_sz;
+} channel_limit_t;
+
+typedef struct {
+	int64_t ticks;
+	MixerEvent cb;
+	void *ctx;
+} mixer_event_t;
+
+struct {
+	uint32_t sample_rate;
+	int num_channels;
+	float divider;
+	float vol;
+
+	int64_t ticks;
+	int num_events;
+	mixer_event_t events[MAX_EVENTS];
+
+	uint8_t *ch_buf_mem;
+	samplebuffer_t ch_buf[MIXER_MAX_CHANNELS];
+	channel_limit_t limits[MIXER_MAX_CHANNELS];
+
+	mixer_channel_t channels[MIXER_MAX_CHANNELS];
+	mixer_fx15_t lvol[MIXER_MAX_CHANNELS];
+	mixer_fx15_t rvol[MIXER_MAX_CHANNELS];
+
+	// Permanent state of the ucode across different executions
+	uint8_t ucode_state[128] __attribute__((aligned(8)));
+
+} Mixer;
+
+/** @brief Count of ticks spent in mixer RSP, used for debugging purposes. */
+int64_t __mixer_profile_rsp = 0;
+
+static inline int mixer_initialized(void) { return Mixer.num_channels != 0; }
+
+void mixer_init(int num_channels) {
+	memset(&Mixer, 0, sizeof(Mixer));
+
+	Mixer.num_channels = num_channels;
+	Mixer.sample_rate = audio_get_frequency();  // actual sample rate obtained via DAC clock
+	assertf(Mixer.sample_rate > 0, "audio_init() must be called before mixer_init()");
+	Mixer.vol = 1.0f;
+
+	for (int ch=0;ch<MIXER_MAX_CHANNELS;ch++) {
+		mixer_ch_set_vol(ch, 1.0f, 1.0f);
+		mixer_ch_set_limits(ch, 16, Mixer.sample_rate, 0);
+	}
+}
+
+static void mixer_init_samplebuffers(void) {
+	// Initialize the samplebuffers. This is done lazily so to allow the
+	// client to configure the limits of the channels.
+	int totsize = 0;
+	int bufsize[MIXER_MAX_CHANNELS];
+
+	for (int i=0;i<Mixer.num_channels;i++) {
+		// Get maximum frequency for this channel: (default: mixer sample rate)
+		int nsamples = Mixer.limits[i].max_frequency;
+		if (!nsamples)
+			nsamples = Mixer.sample_rate;
+
+		// Multiple by maximum byte per sample
+		nsamples *= Mixer.limits[i].max_bits == 8 ? 1 : 2;
+
+		// Calculate buffer size according to number of expected polls per second.
+		bufsize[i] = ROUND_UP((int)ceilf((float)nsamples / (float)MIXER_POLL_PER_SECOND), 8);
+
+		// If we're over the allowed maximum, clamp to it
+		if (Mixer.limits[i].max_buf_sz && bufsize[i] > Mixer.limits[i].max_buf_sz)
+			bufsize[i] = Mixer.limits[i].max_buf_sz;
+
+		assert((bufsize[i] % 8) == 0);
+		totsize += bufsize[i];
+	}
+
+	// Do one large allocations for all sample buffers
+	assert(Mixer.ch_buf_mem == NULL);
+	Mixer.ch_buf_mem = malloc(totsize);
```

Comment:
assert after this malloc to confirm it was successful?

Reply:
Done